### PR TITLE
Wip 10759 - rados cli --striper

### PR DESCRIPTION
**Pull request by sponce**
replaces #4108 and includes fixes to the issues reported by tchaikov

### CODE REVIEW

**Comment by tchaikov** [comment]
the patch changes the default behaviour of `rados`. it looks good other than this.

**Comment by sponce** [comment]
The last addition fixes the change of default behavior. This was just a stupid mistake. Many thanks for spotting it tchaikov !

**Comment by dachary** [comment]
please ignore the false negative from the bog

**Comment by jdurgin** [comment]
Looks good to me. It makes sense to have this in the same tool so the interface stays consistent.

**Comment by athanatos** [comment]
This doesn't seem to want to merge at the moment.

**Comment by athanatos** [comment]
Re-add needs-qa once it merges cleanly.

**Comment by sponce** [comment]
I've rebased it on master. The merge problem was trivial.
I do not think I have the rights to readd needs-qa though. At least I do not find how to achieve it.

**Comment by tchaikov** [comment]
@sponce, `needs-qa` is one of the labels. i can find it at the top-right of this web page, see the gear icon at the right of "Labels"?

**Comment by sponce** [comment]
That's exactly what I suspected : the gear icon is missing for me....

**Comment by liewegas** [comment]
Hrm, needs a rebase again :(

**Comment by tchaikov** [comment]
@sponce ping

**Comment by sponce** [comment]
rebased

**Comment by tchaikov** [comment]
:+1:

**Comment by tchaikov** [comment]
```
./test/erasure-code/test-erasure-code.sh: line 215: 14335 Aborted                 (core dumped) ./rados --block-size=$block_size --pool ecpool put UNALIGNED $dir/ORIGINAL
TEST_alignment_constraints: 227: return 1
run: 47: return 1
```

bt
```
ceph version 0.94-1040-g313143f (313143f560cf11c71bab133c95ecd9a35ff64680)
 1: (ceph::BackTrace::BackTrace(int)+0x2d) [0x4863ff]
 2: /home/jenkins/workspace/ceph/OS/ceph-centos-7-jenkins/src/.libs/lt-rados() [0x485b53]
 3: (()+0xf130) [0x7f1ff66b7130]
 4: (gsignal()+0x37) [0x7f1ff5afd5d7]
 5: (abort()+0x148) [0x7f1ff5afecc8]
 6: (()+0x75e07) [0x7f1ff5b3de07]
 7: (()+0x7d1fd) [0x7f1ff5b451fd]
 8: (librados::NObjectIterator::~NObjectIterator()+0x29) [0x7f1ffa042957]
 9: (__cxa_finalize()+0x9a) [0x7f1ff5b011ba]
 10: (()+0xc807e3) [0x7f1ff77487e3]
```

**Comment by sponce** [comment]
As you mention, this is not related to this patch, first because the --striper option is not used but also because the fact that the --striper option is added is not bringing any new functionality, only an easier interface to existing ones.
Having said that, it is clear that mixing striping and EC pools can create issues of alignment. I'll have to investigate that and make sure one cannot mess up.

**Comment by dachary** [comment]
@sponce could you please repush (after a git commit --amend doing nothing else but changing the commit hash) to trigger the test once more ? It will confirm that it's a false negative. It's the first time I see such an error and I'm not sure it really is a false negative.

**Comment by sponce** [comment]
Sure, done

**Comment by liewegas** [comment]
i think this is causing a lttng double link error,

2015-04-23 13:05:56,745.745 INFO:tasks.util.rados:rados -p rep_pool put REPobject1 /home/ubuntu/cephtest/data/REPobject1
2015-04-23 13:05:56,746.746 INFO:teuthology.orchestra.run.plana05:Running: 'adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage rados -p rep_pool put REPobject1 /home/ubuntu/cephtest/data/REPobject1'
2015-04-23 13:05:56,868.868 INFO:teuthology.orchestra.run.plana05.stderr:LTTng-UST: Error (-17) while registering tracepoint probe. Duplicate registration of tracepoint probes having the same name is not allowed.

**Comment by liewegas** [comment]
yeah, this seems to break lttng.  my guess is that the rados code gets linked twice or something? not sure.  :/

**Comment by tchaikov** [comment]
@sponce indeed, LIBRADOS_TP is linked twice in both librados and libradosstriper, as part of `LIBRADOS_DEPS`.

since libradosstriper is a user facing library, we might want to link rados cli directly against libradosstriper_internal.la which only includes the three sources, and LIBRADOS_DEPS. since libradosstripper is referencing some C++ APIs not exported by librados, like `librados::RadosXattrsIter`. but personally, i'd like link rados cli against librados. as it's a very good example of the rados library.

**Comment by sponce** [comment]
Sorry that I did not have time to look at this so far. I'll try to do something next week.
My problem is that lttng was added after I introduced libradosstriper and that actually created the mess,. I thus have to reverse engineer what was done and how it works to fix it. Anyone having a good understanding of this ? I'm already in contact with Jason Dillaman that did the commit concerning lttng and he gave me a stop-gap solution but I
ll need deeper understanding to make it clean

**Comment by dachary** [comment]
please ignore the false negative from the bot

**Comment by dachary** [comment]
@sponce feel free to re-open as soon as you have it ready

**Comment by sponce** [comment]
ok, no problem

**Comment by dachary** [comment]
@sponce I linked this pull request from the tracker issue as well to be extra sure we won't lose it

**Comment by sponce** [comment]
@dachary I've now fixed the issue (or at least I believe I did), but I'm not able to reopen the pull request. Am I suppose to create a new one or do you have to reopen it yourself ?

**Comment by tchaikov** [comment]
@sponce probably, you can either fetch this pull request and push it again to your `wip-10759` branch, to restore its SHA1 to the one when we closed it. because the hover message reads

> The wip-10759 branch was force-pushed or recreated.

i am not 100% sure about if it works, but it might worth a try, though.

> Am I suppose to create a new one or do you have to reopen it yourself ?

i don't think it must be the one who closed a pull request to reopen it. instead, i guess, the root cause is that you `pushed -f` to `wip-10759` before you tried to reopen it.

**Comment by dachary** [comment]
@sponce like @tchaikov said ;-)

**Comment by dachary** [comment]
@tchaikov I did not realize the one closing the pull request must be the one re-opening it. Good to know.

**Comment by tchaikov** [comment]
> good to know.

ahh, same to me.

**Comment by sponce** [comment]
I think the fact that I've push -f was still creating a mess. I've seen people complaining of this on the web.
Anyway, situation back under control, code ready to be merged. Thanks Loic !

File: src/test/libradosstriper/rados-striper.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Library Public License for more details.
+#
+source test/ceph-helpers.sh
+
+function run() {
+    local dir=$1
+    shift
+
+    export CEPH_MON="127.0.0.1:7112"
```
Comment (dachary) [inline src/test/libradosstriper/rados-striper.sh]: 7112 is taken by src/test/erasure-code/test-erasure-eio.sh

Comment (sponce) [inline src/test/libradosstriper/rados-striper.sh]: And how can I know what to take ? greping shows me that 7101 is also taken, but I have the impression that there is a running counter started at 7100 and I have no clue what is the current value !

Comment (dachary) [inline src/test/libradosstriper/rados-striper.sh]: 7113 seems available, you can take it. I just grep for it and take the next available. Not ideal but I'm not sure how to conveniently do something else.

**Comment by dachary** [comment]
<pre>
run: 25: CEPH_ARGS+='--fsid=383b8946-39b3-4a30-a497-58df59821f4e --auth-supported=none '
run: 26: CEPH_ARGS+='--mon-host=127.0.0.1:7112 '
run: 29: run_mon testdir/rados-striper a
run_mon: 280: local dir=testdir/rados-striper
run_mon: 281: shift
run_mon: 282: local id=a
run_mon: 283: shift
run_mon: 284: local data=testdir/rados-striper/a
run_mon: 287: ceph-mon --id a --mkfs --mon-data=testdir/rados-striper/a --run-dir=testdir/rados-striper
mkdir(testdir/rados-striper/a) : (2) No such file or directory
run_mon: 291: return 1
</pre>

because you need to **setup $dir || return 1** at the beginning of the test function and finish with **teardown $dir**. It used to be different, which is why your function does not have it.

**Comment by dachary** [comment]
the last commit has a civetweb module update. Could you please take the opportunity of the rebase to have a title prefixed with **tests:** ? And a it's better if the description does not include **latest** but rather the SHA of the corresponding commit for clarity. Thanks !

**Comment by sponce** [comment]
done. I had no clue for the sha, so I've put something more descriptive

**Comment by dachary** [comment]
looks ready for qa, when the spurious civetweb module update is removed

**Comment by sponce** [comment]
Ok, now I got your point on civeweb module and dropped the module update (no clue where it came from).

**Comment by tchaikov** [comment]
@sponce would you rebase again =) thanks !

**Comment by sponce** [comment]
done

**Comment by athanatos** [comment]
Needs to be rebased again.

**Comment by sponce** [comment]
Done again. Could it be rather merged so that I do not have to rebase it regularly ?

File: src/test/libradosstriper/rados-striper.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# it under the terms of the GNU Library Public License as published by
+# the Free Software Foundation; either version 2, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Library Public License for more details.
+#
+source test/ceph-helpers.sh
```
Comment (tchaikov) [inline src/test/libradosstriper/rados-striper.sh]: @sponce might want to `source ../qa/workunits/ceph-helpers.sh` instead.

**Comment by tchaikov** [comment]
```
./test/libradosstriper/rados-striper.sh: line 17: test/ceph-helpers.sh: No such file or directory
./test/libradosstriper/rados-striper.sh: line 95: main: command not found
```

@sponce sorry to keep you rebasing from time to time. =(

will include your pull request in my next qa run, once the issue above is taken care of.

**Comment by sponce** [comment]
No worry, I can do one more rebase :-)
But we are all losing time. See this last problem : test/ceph-helpers.sh was moved by Loic 3 weeks ago, creating another iteration both for you and me.

**Comment by tchaikov** [comment]
tested at
- http://pulpito.ceph.com/ubuntu-2015-06-29_18:58:24-rados-wip-kefu-testing---basic-multi/
- http://pulpito.ceph.com/ubuntu-2015-06-30_20:30:57-rados-wip-kefu-testing---basic-multi/

the failed ones are tracked at http://tracker.ceph.com/issues/12192, which are also spotted on the master branch.

**Pull request merged**